src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs

```diff
@@ -112,14 +112,63 @@ await InnerProcessMessages()
 
                     var nativeMessageId = Path.GetFileNameWithoutExtension(filePath).Replace(".metadata", "");
 
-                    var transaction = GetTransaction();
+                    await concurrencyLimiter.WaitAsync(cancellationToken)
+                        .ConfigureAwait(false);
 
-                    transaction.BeginTransaction(filePath);
+                    var transaction = GetTransaction();
 
-                    await concurrencyLimiter.WaitAsync(cancellationToken)
+                    var ableToLockFile = await transaction.BeginTransaction(filePath)
                         .ConfigureAwait(false);
 
-                    InnerProcessFile(transaction, nativeMessageId).Ignore();
+                    if (!ableToLockFile)
+                    {
+                        concurrencyLimiter.Release();
+                        continue;
+                    }
+
+                    ProcessFile(transaction, nativeMessageId)
+                        .ContinueWith(async t =>
+                        {
+                            try
+                            {
+                                if (t.Exception == null)
+                                {
+                                    receiveCircuitBreaker.Success();
+                                }
+                                else
+                                {
+                                    var baseEx = t.Exception.GetBaseException();
+
+                                    if (!(baseEx is OperationCanceledException))
+                                    {
+                                        await receiveCircuitBreaker.Failure(baseEx)
+                                            .ConfigureAwait(false);
+                                    }
+
+                                    transaction.Rollback();
+
+                                }
+
+                                var wasCommitted = await transaction.Complete()
+                                    .ConfigureAwait(false);
+
+                                if (wasCommitted)
+                                {
+                                    File.Delete(Path.Combine(bodyDir, nativeMessageId + BodyFileSuffix));
+                                }
+                            }
+                            catch (Exception ex)
+                            {
+                                await receiveCircuitBreaker.Failure(ex)
+                                    .ConfigureAwait(false);
+                            }
+                            finally
+                            {
+                                concurrencyLimiter.Release();
+                            }
+                        }, CancellationToken.None)
```

Comment:
I had the feeling that it could make sense to move the transaction completion "closer" to the completion of the processing task but yeah, the continuation is fairly long and async as well, so I guess it's indeed better not to do so.